### ISSUE TITLE
false positive for `unused_parens` with binary op if rhs needs parens

### ISSUE THREAD

**Issue by lukas-code** [rust-lang/rust#110251]
I tried this *very normal* code: [link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=679c27403c72186bed74c87b809fc4c9)

```rust
#![allow(unreachable_code)]

fn main() {
    if (() = return) {}
    if ((..{}) == ..{}) {}
}
```

The current output is the following, but removing any of the parens is an error:

```text
   Compiling playground v0.0.1 (/playground)
warning: unnecessary parentheses around `if` condition
 --> src/main.rs:4:8
  |
4 |     if (() = return) {}
  |        ^           ^
  |
  = note: `#[warn(unused_parens)]` on by default
help: remove these parentheses
  |
4 -     if (() = return) {}
4 +     if () = return {}
  |

warning: unnecessary parentheses around `if` condition
 --> src/main.rs:5:8
  |
5 |     if ((..{}) == ..{}) {}
  |        ^              ^
  |
help: remove these parentheses
  |
5 -     if ((..{}) == ..{}) {}
5 +     if (..{}) == ..{} {}
  |

warning: `playground` (bin "playground") generated 2 warnings
    Finished dev [unoptimized + debuginfo] target(s) in 0.99s
     Running `target/debug/playground`
```

This also gets `cargo fix`ed incorrectly.

This need to change to look at the rhs of binops: https://github.com/rust-lang/rust/blob/661b33f5247debc4e0cd948caa388997e18e9cb8/compiler/rustc_lint/src/unused.rs#L591-L601

@rustbot label T-compiler A-lint A-diagnostics
